### PR TITLE
Text Editing Model Refactor

### PR DESCRIPTION
**Pull request by justinmc**
This PR aims to improve the separation of concerns of our text editing classes.

Design doc: https://docs.google.com/document/d/1lkpg4dfVZif9NJ_dDQ01QkwQC3OqzhUFahhjgv_rcbM/edit#

Related: https://github.com/flutter/flutter/issues/80226

### Summary

- TextEditingValue and TextSelection now contain simple methods like extendTo, deleteTo, etc. for manipulating themselves.  The logic was abstracted from RenderEditable.
- TextEditingActionTarget now contains the more specific text manipulation methods, which often call out to TextEditingValue (e.g. extendSelectionRightByLine, deleteToStart, etc.).  These were also abstracted from RenderEditable.
- TextMetrics is a new abstract class, implemented by RenderEditable, that exposes an API for getting text position information only.  It's sometimes used by the TextEditingActionTarget methods.
- Overall result: RenderEditable shrinks by ~2,000 lines of logic unrelated to rendering.  All of that logic no longer has access to data from multiple frames, no write access beyond the TextEditingValue.

### Reviewing
Sorry this is such a huge diff!  It's all moving code around and refactoring it; very little totally new.  You can skim most of it because of this.  I tried to leave helpful comments where big chunks of code were moved.  Let me know if I can explain anything.

My commits are also pretty messy as I explored ideas and changed approaches 😁

### CODE REVIEW

File: packages/flutter/lib/src/services/text_input.dart
Comment (justinmc) [inline packages/flutter/lib/src/services/text_input.dart:658]: TextEditingValue was moved to its own file in order to not pollute this one, now that it's much larger.

File: packages/flutter/lib/src/widgets/text_editing_action.dart
Comment (justinmc) [inline packages/flutter/lib/src/widgets/text_editing_action.dart:472]: This is probably one of the big things I still need to do.  I think that doing the assert in RenderEditable as I described in this comment is doable.  I kind of wish we didn't need this assertion at all, though.

File: packages/flutter/lib/src/rendering/editable.dart
Comment (justinmc) [inline packages/flutter/lib/src/rendering/editable.dart:80]: This file's diff is really hard to read, but basically what happened was:

  * Lots of methods moved to TextEditingAction and TextEditingValue.
  * Some alphabetizing/reordering may have happened...
  * TextMetrics was implemented.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (justinmc) [inline packages/flutter/test/widgets/editable_text_test.dart:4417]: Previously, we were usually resetting the affinity when manipulating the selection.  I updated everything to preserve the affinity, or explicitly set it when it makes sense (like moving the selection to the start should be upstream).  I updated these tests to match.

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (justinmc) [inline packages/flutter/lib/src/services/text_editing_value.dart:1]: Are we ok with growing the API of TextEditingValue like this?

File: packages/flutter/lib/src/rendering/editable.dart
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/editable.dart:544]: This is where the one frame delay I mentioned in the doc. the _textPainter.text may be out of date from the text editing value. The caretOffset, or any other metrics render editable collected from the textpainter, may be out of date for one frame. I added a hack in _setSelection to get around this problem, but I am not sure whether there will be other corner case

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (chunhtai) [inline packages/flutter/lib/src/services/text_editing_value.dart:123]: I have written similar typographic methods on the engine side
https://github.com/flutter/engine/blob/e0ebbb226f38e552289e380324943ec67126d325/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L1192
and
https://github.com/flutter/engine/blob/e0ebbb226f38e552289e380324943ec67126d325/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm#L437

I hope at some point we can use the API from skia to calculate code unit and avoid duplicating the logic in a bunch of places...

**Review by chunhtai — COMMENTED** [review]
Is it possible to separate this to smaller code chunks?

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_editing_value.dart:191]: Ah I see `TextMetrics` acts as a tokenizer here. But I don't think this is right: `textMetrics` may have different text than the `text` passed in.

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_editing_value.dart:323]: It's not clear to me what the differences are between expand and extend from reading the documentation.

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_editing_value.dart:406]: Why the `copyWith`? Is it for keeping the original `isDirectional`? (I'm not sure what `isDirectional` is for, none of our platform text input plugins set that to true it seems).

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_editing_value.dart:27]: Maybe we should have a `TextEditingValue replace(TextRange, String)` method like Android does? One of the biggest pain points from my personal experience is input manipulation is tricky when you have to take the composing range and selection range into account when you just want to change the text.

File: packages/flutter/lib/src/widgets/text_editing_action_target.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_editing_action_target.dart:22]: This class seems to have way too many methods that are very specific. Hopefully we can distill them down to a handful as basis.

**Review by LongCatIsLooong — COMMENTED** [review]
Unfortunately I don't think this fixes https://github.com/flutter/flutter/issues/80226 yet (if the issue includes `deleteByWord` and `deleteByLine`, etc). But this looks like a good step towards that!

File: packages/flutter/lib/src/rendering/editable.dart
Comment (justinmc) [reply packages/flutter/lib/src/rendering/editable.dart:544]: Ah I see, thanks for the explanation.  I'll take another look at what we can do to mitigate that problem after this is merged.

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing_value.dart:177]: I actually had already moved this to TextEditingActionTarget as a private instance method that uses TextMetrics.getWordPosition, but I had forgotten to delete this static method here.  Thanks for catching this.  Let me know if you like it on TextEditingActionTarget or if you think it would be better on TextMetrics.

Regarding RTL, you're right, I think we have a lot of missing support for this.  I've created an issue (https://github.com/flutter/flutter/issues/88126) and I'll add TODOs for it here.

File: packages/flutter/lib/src/services/text_metrics.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_metrics.dart]: Good catch, I tried to convert most things to use a TextPosition but missed these.

File: packages/flutter/lib/src/widgets/text_editing_action_target.dart
Comment (justinmc) [reply packages/flutter/lib/src/widgets/text_editing_action_target.dart:22]: That seems like it'll be good to me, I'll ping you for more details later.

I thought about putting all of the logic that is currently in these methods directly into the Action handlers instead, which would also work.  However I worry about encouraging us/developers to use Actions when plain ol' methods are easier to work with and understand.

I have simplified the methods in TextEditingValue per your recommendation.  So TextEditingValue contains simple methods like moveSelectionTo, while TextEditingActionTarget contains the higher level methods like moveSelectionForwardByWord.

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing_value.dart]: You're right, same problem as ever in that the TextEditingValue and TextPainter's text can be out of sync.  I'll follow up on this.

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing_value.dart:303]: I knew that was going to be confusing.  I'll add some docs about it in both expandSelectionTo and extendSelectionTo.

Basically, "extend" moves the extentOffset while holding the baseOffset fixed, but "expand" grows the selection to include the given position.  If you try cmd+shift+left/right on a Mac, that's expand, while option+shift+up/down is extend.

Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing_value.dart:303]: I've improved the docs: https://github.com/flutter/flutter/pull/86736/commits/0675ec4ad718f1202101edf3d0c1d6167a4bd9ce

Would it be more intuitive if I renamed "extend" to "pivot"?

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing_value.dart:397]: Yes, I didn't want to mess with that as I also wasn't sure where it matters.

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing_value.dart:27]: I'd like to hold off on this to avoid increasing the size of this PR any more if that's ok with you, but I really like the idea.  Lots of the existing code in the delete methods here spend a lot of lines getting the composing region right.  It would be cool to just remove the relevant text from the string and then call `replace` and let it figure out composing and selection.  I'll keep an eye out for other opportunities for this.

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing_value.dart:227]: I've changed it to say "inclusive" but tell me if I'm wrong here...

`wo|rd`

deleteTo(0) => `|rd`
deleteTo(4) => `wo|`

Is that right?  I think it confuses me because I'm thinking of the indices as the positions between the characters, not the characters themselves.

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing_value.dart]: Done here: b3864a9

I feel a little uneasy about adding new methods to TextSelection just like I did originally for TextEditingValue, but maybe there's nothing to worry about.  These methods are pretty fundamental.

**Comment by justinmc** [comment]
@chunhtai @LongCatIsLooong Ready for re-review.

About splitting this up into smaller PRs, unfortunately I think it's more work than it's worth at this point.  Sorry, I know the diff is enormous.  I tried to leave comments where things were moved and refactored so it's not like reviewing brand new code.

About the cross-frame problem, I may take a look at fixing the root part in another PR.  At least all of these methods in this PR have access to one source of truth only now.

Lastly, the only failure now is the typical SVG customer test, which I have fixed before by squashing all the commits.  Hopefully it just goes away before I have to do that.

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_editing_value.dart:242]: There seems to be code duplication, 243-260 looks very similar to 262 - 281. This part of the implementation will probably be cleaner if we first figure out the normalized `TextRange` to delete, and then use "start" and "end" instead of "index" and "selection.extentOffset".

File: packages/flutter/lib/src/services/text_metrics.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_metrics.dart:27]: `_plainText` is private.

File: packages/flutter/lib/src/services/text_editing.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_editing.dart:168]: This is `end`. and `lowerOffset` is `start`.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_editing.dart:168]: nit: the implementation is a bit difficult to read.

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_editing_value.dart]: Yeah that's my understanding as well.

File: packages/flutter/lib/src/services/text_editing.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing.dart]: Agreed.  I cleaned this up a bit and added some comments.

**Comment by justinmc** [comment]
@LongCatIsLooong Ready for re-review.

File: packages/flutter/lib/src/services/text_metrics.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_metrics.dart:54]: IMO this is going to be error prone. `TextMetrics`'s layout information is highly dependent on its text, so being able to provide a potentially different String is a bit confusing. 

> enforce what text it's using

Yeah but we can't layout a plain `String`. It has to be styled and you need other inputs for instance the incoming box constraints. Also other methods in the interface don't seem to take a String? All `TextMetrics` implementations are likely going to have states.

And for `RenderEditable.getLineAtOffset`, getting the length of the obscured text should be an implementation detail instead of being exposed as a parameter I think.

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_editing_value.dart:27]: Yeah the android `Editable` implementation implements most common text manipulation methods in terms of `replace`, since it's `replace` implementation takes care of the selection and composition.

File: packages/flutter/lib/src/rendering/editable.dart
Comment (justinmc) [reply packages/flutter/lib/src/rendering/editable.dart:540]: Yup good catch.

File: packages/flutter/lib/src/services/text_editing.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing.dart]: Good point, which as we've discussed doesn't seem to be relevant to this.  I'll swap in `fromPosition` and get rid of this method.

File: packages/flutter/lib/src/services/text_metrics.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_metrics.dart:16]: It looks like it's used in 3 different files now unfortunately (RenderEditable, TextEditingValue, TextEditingActionTarget).  Probably not worth it to duplicate the method three times.

I see @gspencergoog's comment about this:

```
// TODO(gspencergoog): replace when we expose this ICU information.
```

Any idea if we still have plans to expose that, or if there's a cleaner way for me to expose this one?

File: packages/flutter/lib/src/services/text_metrics.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_metrics.dart:54]: You're right, this parameter makes the multiple sources of truth thing more confusing, not less.  What we need to do is ensure that the instance of TextMetrics etc. are consistent with the state they're operating on.  I'll remove the parameter and use `_plainText`.

File: packages/flutter/lib/src/services/text_metrics.dart
Comment (gspencergoog) [reply packages/flutter/lib/src/services/text_metrics.dart:16]: I don't know if we plan to expose it, but I think the problem was that it adds to executable size too much to add the tables in that ICU uses.

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing_value.dart]: Ah good catch.  I remember talking to the Dart team about this but can't remember the details.  This seems to work with my testing.

I'll refactor this to use CharacterRange, move it into test_editing_action_target.dart, and make it `visibleForTesting` only.

Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing_value.dart]: Here's that change: bd27952

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing_value.dart]: I ended up removing selectAll and doing it all in TextEditingActionTarget. 6e74a3d

File: packages/flutter/lib/src/services/text_editing_value.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_editing_value.dart]: I investigated adding `replace` to this PR again, but it is making this PR even bigger and more confusing.  I'll follow up in another PR based on my branch https://github.com/justinmc/flutter/tree/text-editing-value-replace.  In the meantime I've removed all changes to TextEditingValue.  `deleteTo` and related stuff is not private inside of TextEditingActionTarget.  In the new PR, I'll replace that with calls to `replace`.

**Review by LongCatIsLooong — APPROVED** [review]
LGTM. I wonder if there's a better name for `TextMetrics`. `TextLayout`?

**Pull request merged**
